sae_lens/sae.py

```diff
@@ -491,9 +488,39 @@ def forward(
 
         return self.hook_sae_output(sae_out)
 
+    def encode(
+        self, x: torch.Tensor, latents: torch.Tensor | None = None
+    ) -> torch.Tensor:
+        """
+        Calculate SAE latents from inputs. Includes optional `latents` argument to only calculate a subset. Note that
+        this won't make sense for topk SAEs, because we need to compute all hidden values to apply the topk masking.
+        """
+        if self.cfg.activation_fn_str == "topk":
+            assert (
+                latents is None
+            ), "Computing a slice of SAE hidden values doesn't make sense in topk SAEs."
+
+        return {
+            "standard": self.encode_standard,
+            "gated": self.encode_gated,
+            "jumprelu": self.encode_jumprelu,
+        }[self.cfg.architecture](x, latents)
+
     def encode_gated(
-        self, x: Float[torch.Tensor, "... d_in"]
+        self,
+        x: Float[torch.Tensor, "... d_in"],
+        latents: torch.Tensor | None = None,
     ) -> Float[torch.Tensor, "... d_sae"]:
+        """
+        Computes the latent values of the Sparse Autoencoder (SAE) using a gated architecture. The activation values are
+        computed as the product of the masking term & the post-activation function magnitude term:
+
+            1[(x - b_dec) @ W_gate + b_gate > 0] * activation_fn((x - b_dec) @ W_enc + b_enc)
+
+        The `latents` argument allows for the computation of a specific subset of the hidden values. If `latents` is not
+        provided, all latent values will be computed.
+        """
+        latents_tensor = torch.arange(self.cfg.d_sae) if latents is None else latents
```

Comment:
The previous implementation you had, `latents_slice = slice(None) if latents is None else torch.tensor(latents)`, seems better to me since this new version will create a new tensor of size `d_sae` on every SAE forward pass when not selecting specific latents. This would likely reduce performance for most users, which seems counter-productive since this PR is just meant to be a performance improvement if I understand the goal correctly. Wouldn't the old implementation have worked fine just adding `torch.Tensor` to the type of `latents`? e.g. `latents: Iterable[int] | torch.Tensor | None = None`